cli/command/formatter/sorter.go

```diff
@@ -0,0 +1,321 @@
+
+			if diField.Type.Kind() == reflect.Ptr {
+				// keep nil value before non-nil value
+				// Note: for Less() parameters, i is larger than j, e.g. i=1, j=0
+				// if djFieldValue(data[0]) is nil, return false to keep their positions
+				// or if djFieldValue(data[0]) isn't nil but diFieldValue(data[1]) is nil, return true to swap
+				if djFieldValue.IsNil() {
+					return false
+				} else if diFieldValue.IsNil() {
+					return true
+				}
+
+				diCurType = diField.Type.Elem()
+				diCurValue = diFieldValue.Elem()
+
+				djCurType = djField.Type.Elem()
+				djCurValue = djFieldValue.Elem()
+			} else {
+				diCurType = diField.Type
+				diCurValue = diFieldValue
+
+				djCurType = djField.Type
+				djCurValue = djFieldValue
+			}
+		}
+
+		res := 0
+		switch diCurType.Kind() {
+		case reflect.String:
+			res = lessForString(diCurValue, djCurValue)
+		case reflect.Bool:
+			res = lessForBool(diCurValue, djCurValue)
+		case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+			res = lessForInt(diCurValue, djCurValue)
+		case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+			res = lessForUint(diCurValue, djCurValue)
+		case reflect.Float32, reflect.Float64:
+			res = lessForFloat(diCurValue, djCurValue)
+		default:
+			switch diCurType {
+			case reflect.TypeOf(time.Time{}):
+				res = lessForTime(diCurValue, djCurValue)
+			default:
+				// panic(fmt.Sprintf("genericStructSorter: comparison of %v are not supported", diCurType.Kind()))
+				// TODO: find a reasonable logic
+				// if it's unsupported field, keep original positions
+				return false
+			}
+		}
+
+		if res != 0 {
+			if orderBy.ordering == ascendantKey {
+				return res == -1
+			}
+			return res == 1
+		}
+	}
+
+	// equal values, return false to keep their postions
+	return false
+}
+
+func lessForString(i, j reflect.Value) int {
+	vi, vj := i.String(), j.String()
+	if vi == vj {
+		return 0
+	}
+	if vi < vj {
+		return -1
+	}
+	return 1
+}
+
+func lessForBool(i, j reflect.Value) int {
+	vi, vj := i.Bool(), j.Bool()
+	if vi == vj {
+		return 0
+	}
+	if !vi && vj {
+		return -1
+	}
+	return 1
+}
+
+func lessForInt(i, j reflect.Value) (ret int) {
+	vi, vj := i.Int(), j.Int()
+	if vi == vj {
+		return 0
+	}
+	if vi < vj {
+		return -1
+	}
+	return 1
+}
+
+func lessForUint(i, j reflect.Value) int {
+	vi, vj := i.Uint(), j.Uint()
+	if vi == vj {
+		return 0
+	}
+	if vi < vj {
+		return -1
+	}
+	return 1
+}
+
+func lessForFloat(i, j reflect.Value) int {
+	vi, vj := i.Float(), j.Float()
+	if vi == vj {
+		return 0
+	}
+	if vi < vj {
+		return -1
+	}
+	return 1
+}
+
+func lessForTime(i, j reflect.Value) int {
+	vi, vj := i.Interface().(time.Time), j.Interface().(time.Time)
+	if vi.Equal(vj) {
+		return 0
+	}
+	if vi.Before(vj) {
+		return -1
+	}
+	return 1
+}
+
+func getSortableFields(dataType reflect.Type, whitelist map[string]string, parent fieldIndex, fields map[string]fieldIndex) error {
```

Comment:
Because this is a recursive function, we need to pass `fields` from top to bottom, and all the levels will filling the same `fields` param.